### PR TITLE
Make Kdyby/Events optional

### PR DESCRIPTION
**Pull request by enumag**
#239 

I wanted to add some tests but unfortunately I was not even able to run the current tests - PHP just crashed. :-( Which means this is highly experimental.

### CODE REVIEW

**Comment by enumag** [comment]
Travis is failing because of some incompatibility in dev Nette.

**Comment by enumag** [comment]
Failing because of https://github.com/Kdyby/Console/pull/54.

File: src/Kdyby/Doctrine/DI/OrmExtension.php [hunk 1]
```diff
@@ -630,25 +649,6 @@ protected function processCache($cache, $suffix)
...
-		if ($eventsExt === NULL) {
-			throw new Nette\Utils\AssertionException('Please register the required Kdyby\Events\DI\EventsExtension to Compiler.');
-		}
-
-		$this->processRepositories();
-	}
-
-
-
-	protected function processRepositories()
```
Comment (fprochazka) [inline src/Kdyby/Doctrine/DI/OrmExtension.php]: I would like to keep this as a separate method

Comment (enumag) [inline src/Kdyby/Doctrine/DI/OrmExtension.php]: No problem. It's just that beforeCompile will only call this method and nothing else which seems a bit odd. But sure. :-)

**Comment by fprochazka** [comment]
I've cherrypicked and fixed few missing occurences of `->class`, https://github.com/Kdyby/Doctrine/commit/3e8bd5deb6c5aab514c62c228a14fb86f9846ce1. Thanks!

**Comment by enumag** [comment]
Travis finally passed after rebase. Which means I at least didn't break any tests. I'm still unable to run the tests on localhost though so writing a new test is not easy.

**Comment by enumag** [comment]
Finally managed to run the tests on localhost by using `-p php` (from .travis.yml).

In the end I didn't write any new tests, just changed the old ones little bit:

1. Replaced `EventsExtension` with a simple service `Doctrine\Common\EventManager` for all tests.
2. Identified tests broken in step 1 - Events.compatibility.phpt and TargetEntityMapping.phpt.
3. These tests do require Kdyby/Events so I refactored them to use different configuration.

**Comment by enumag** [comment]
@fprochazka With the tests updated I'm satisfied with this PR as it is. Can you review the changes please? :-)

**Comment by enumag** [comment]
Refactored Kdyby/Events detection since you recently wanted a method instead of a property in similar case in Kdyby/Console.

File: tests/KdybyTests/Doctrine/ORMTestCase.php [hunk 2]
```diff
@@ -64,6 +69,12 @@ protected function createMemoryManager()
 
 
 
+	protected function configure(Configurator $config)
```
Comment (fprochazka) [inline tests/KdybyTests/Doctrine/ORMTestCase.php]: this is not used anywhere

**Comment by fprochazka** [comment]
@enumag Thank you, merged! I've made some minor changes

```patch
diff --git a/src/Kdyby/Doctrine/DI/OrmExtension.php b/src/Kdyby/Doctrine/DI/OrmExtension.php
index 7155fd4..0da259f 100644
--- a/src/Kdyby/Doctrine/DI/OrmExtension.php
+++ b/src/Kdyby/Doctrine/DI/OrmExtension.php
@@ -351,24 +351,26 @@ class OrmExtension extends Nette\DI\CompilerExtension
                        $this->targetEntityMappings = Nette\Utils\Arrays::mergeTree($this->targetEntityMappings, $config['targetEntityMappings']);
                }
 
-               $entityManagerArguments = [
-                       $connectionService = $this->processConnection($name, $defaults, $isDefault),
-                       $this->prefix('@' . $name . '.ormConfiguration'),
-               ];
-
                if ($this->isKdybyEventsPresent()) {
                        $builder->addDefinition($this->prefix($name . '.evm'))
                                ->setClass('Kdyby\Events\NamespacedEventManager', [Kdyby\Doctrine\Events::NS . '::'])
                                ->addSetup('$dispatchGlobalEvents', [TRUE]) // for BC
                                ->setAutowired(FALSE);
 
-                       $entityManagerArguments[] = $this->prefix('@' . $name . '.evm');
+               } else {
+                       $builder->addDefinition($this->prefix($name . '.evm'))
+                               ->setClass('Doctrine\Common\EventManager')
+                               ->setAutowired(FALSE);
                }
 
                // entity manager
                $entityManager = $builder->addDefinition($managerServiceId = $this->prefix($name . '.entityManager'))
                        ->setClass('Kdyby\Doctrine\EntityManager')
-                       ->setFactory('Kdyby\Doctrine\EntityManager::create', $entityManagerArguments)
+                       ->setFactory('Kdyby\Doctrine\EntityManager::create', [
+                               $connectionService = $this->processConnection($name, $defaults, $isDefault),
+                               $this->prefix('@' . $name . '.ormConfiguration'),
+                               $this->prefix('@' . $name . '.evm'),
+                       ])
                        ->addTag(self::TAG_ENTITY_MANAGER)
                        ->addTag('kdyby.doctrine.entityManager')
                        ->setAutowired($isDefault)
@@ -523,19 +525,13 @@ class OrmExtension extends Nette\DI\CompilerExtension
 
                // connection
                $options = array_diff_key($config, array_flip(['types', 'resultCache', 'connection', 'logging']));
-
-               $connectionArguments = [
-                       $options,
-                       $this->prefix('@' . $name . '.dbalConfiguration'),
-               ];
-
-               if ($this->isKdybyEventsPresent()) {
-                       $connectionArguments[] = $this->prefix('@' . $name . '.evm');
-               }
-
                $connection = $builder->addDefinition($connectionServiceId = $this->prefix($name . '.connection'))
                        ->setClass('Kdyby\Doctrine\Connection')
-                       ->setFactory('Kdyby\Doctrine\Connection::create', $connectionArguments)
+                       ->setFactory('Kdyby\Doctrine\Connection::create', [
+                               $options,
+                               $this->prefix('@' . $name . '.dbalConfiguration'),
+                               $this->prefix('@' . $name . '.evm'),
+                       ])
                        ->addSetup('setSchemaTypes', [$schemaTypes])
                        ->addSetup('setDbalTypes', [$dbalTypes])
                        ->addTag(self::TAG_CONNECTION)
diff --git a/tests/KdybyTests/Doctrine/ORMTestCase.php b/tests/KdybyTests/Doctrine/ORMTestCase.php
index 0d6a521..b6d0f4f 100644
--- a/tests/KdybyTests/Doctrine/ORMTestCase.php
+++ b/tests/KdybyTests/Doctrine/ORMTestCase.php
@@ -69,12 +69,6 @@ abstract class ORMTestCase extends Tester\TestCase
 
 
 
-       protected function configure(Configurator $config)
-       {
-       }
-
-
-
        /**
         * @param string $className
         * @param array $props
diff --git a/tests/KdybyTests/nette-reset.neon b/tests/KdybyTests/nette-reset.neon
index 0e29325..cc18417 100644
--- a/tests/KdybyTests/nette-reset.neon
+++ b/tests/KdybyTests/nette-reset.neon
@@ -31,5 +31,3 @@ session:
 services:
        cacheStorage:
                class: Nette\Caching\Storages\MemoryStorage
-       evm:
-               class: Doctrine\Common\EventManager
```

**Comment by enumag** [comment]
If I'm not mistaken those minor changes break what I intended - I need the evm to be autowired, not use some dummy replacement

**Comment by fprochazka** [comment]
I know, but I don't want the users to have to configure some magic service, that should be configured by Kdyby.
If you wanna provide custom implementation of `Doctrine\Common\EventManager`, you can override the service for the EM.

Now you can
- use this with Kdyby/Events and it works
- use this without Kdyby/Events and it works
- use this without Kdyby/Events, with custom implementation of `EventManager` and it works.

IMHO it's cleaner that way and it makes sense that the third requires you to touch the `evm` service in config.

**Comment by enumag** [comment]
I can't agree. I have an providing a simple ContainerAwareEventManager and wanted to use it in Kdyby Doctrine. However my extension should not be aware of Kdyby/Doctrine at all (it's independent) - you're forcing me to specifically look for OrmExtension and hack it.

**Comment by enumag** [comment]
How about adding the evm service only if there is no autowired one?

**Comment by fprochazka** [comment]
First of all 

> I don't want the users to have to configure some magic service, that should be configured by Kdyby.

I hope you understand that. I wanna avoid this at all costs.

By just simply accepting what you've proposed, It would be simple for you, but I'd cripple the UX of the lib for everyone else.

----

About your argument - I don't consider it hacking. But I understand why you don't like it. 

What about adding another case - if Kdyby/Events is not registered and a service of type `Doctrine\Common\EventManager` can be found in config, the `evm` service will be an alias to that service.

**Comment by enumag** [comment]
How about adding the evm service only if there is no autowired one?

**Comment by fprochazka** [comment]
Imho it would be better to do something like

```
$evm = $builder->addDefinition($this->prefix($name . '.evm'))
	->setClass('Doctrine\Common\EventManager')
	->setAutowired(FALSE);

if ($customEvm = $builder->getByType('Doctrine\Common\EventManager')) {
	$evm->setFactory('@' . $customEvm);
}
```

but that's a detail :)

**Comment by enumag** [comment]
@fprochazka Yeah. This is fine by me. :-)

**Comment by fprochazka** [comment]
My point is that you don't have to introduce more if's in the rest of the code :)

**Comment by enumag** [comment]
I'm aware... I didn't like the ifs in my PR either. :-)

**Comment by fprochazka** [comment]
I'm lazy to write a test for it right now. You mind doing it and sending a PR? :bowtie:

**Comment by enumag** [comment]
Well... You're well aware that I won't refuse. :-D

**Pull request closed without merging**